udtf.go

```diff
@@ -0,0 +1,493 @@
+	if err != nil {
+		errstr := C.CString(err.Error())
+		defer C.free(unsafe.Pointer(errstr))
+		C.duckdb_bind_set_error(info, errstr)
+		return
+	}
+
+	columns := instance.Columns()
+
+	instanceData := tableFunctionData{
+		fun:        instance,
+		projection: make([]int, len(columns)),
+	}
+
+	for i, v := range columns {
+		dt := v.T.logicalType()
+		defer C.duckdb_destroy_logical_type(&dt)
+		colName := C.CString(v.Name)
+		defer C.free(unsafe.Pointer(colName))
+		C.duckdb_bind_add_result_column(info, colName, dt)
+
+		instanceData.projection[i] = -1
+	}
+
+	cardinality := instance.Cardinality()
+	if cardinality != nil {
+		C.duckdb_bind_set_cardinality(info, C.idx_t(cardinality.Cardinality), C.bool(cardinality.IsExact))
+	}
+
+	pinnedInstanceData := pinnedValue[tableFunctionData]{
+		pinner: &runtime.Pinner{},
+		value:  instanceData,
+	}
+
+	handle := cgo.NewHandle(pinnedInstanceData)
+	pinnedInstanceData.pinner.Pin(&handle)
+	C.duckdb_bind_set_bind_data(info, unsafe.Pointer(&handle), C.duckdb_delete_callback_t(C.udf_destroy_data))
+}
+
+//export udf_init
+func udf_init(info C.duckdb_init_info) {
+	h := *(*cgo.Handle)(C.duckdb_init_get_bind_data(info))
+	instance := h.Value().(pinnedValue[tableFunctionData]).value
+
+	columnCount := C.duckdb_init_get_column_count(info)
+	for i := C.idx_t(0); i < columnCount; i++ {
+		srcPos := int(C.duckdb_init_get_column_index(info, i))
+		instance.projection[srcPos] = int(i)
+	}
+	instance.fun.(tableSource).Init()
+}
+
+//export udf_init_threaded
+func udf_init_threaded(info C.duckdb_init_info) {
+	h := *(*cgo.Handle)(C.duckdb_init_get_bind_data(info))
+	instance := h.Value().(pinnedValue[tableFunctionData]).value
+
+	columnCount := C.duckdb_init_get_column_count(info)
+	for i := C.idx_t(0); i < columnCount; i++ {
+		srcPos := int(C.duckdb_init_get_column_index(info, i))
+		instance.projection[srcPos] = int(i)
+	}
+
+	initData := instance.fun.(threadedTableSource).Init()
+	C.duckdb_init_set_max_threads(info, C.idx_t(initData.MaxThreads))
+}
+
+//export udf_local_init
+func udf_local_init(info C.duckdb_init_info) {
+	h := *(*cgo.Handle)(C.duckdb_init_get_bind_data(info))
+	instance := h.Value().(pinnedValue[tableFunctionData]).value
+	localState := pinnedValue[any]{
+		pinner: &runtime.Pinner{},
+		value:  instance.fun.(threadedTableSource).NewLocalState(),
+	}
+	handle := cgo.NewHandle(localState)
+	localState.pinner.Pin(&handle)
+	C.duckdb_init_set_init_data(info, unsafe.Pointer(&handle), C.duckdb_delete_callback_t(C.udf_destroy_data))
+}
+
+//export udf_row_callback
+func udf_row_callback(info C.duckdb_function_info, output C.duckdb_data_chunk) {
+	h := *(*cgo.Handle)(C.duckdb_function_get_bind_data(info))
+	instance := h.Value().(pinnedValue[tableFunctionData]).value
+
+	var chunk DataChunk
+	// false, maybe this needs to be true but I don't think we need to read the validity mask
+	// It should be set after executing the function, but never read.
+	err := chunk.initFromDuckDataChunk(output, false)
```

Comment:
It should be `true`; otherwise, setting values can cause uninitialized memory crashes.